pytensor/scalar/loop.py

```diff
@@ -0,0 +1,388 @@
+    def perform(self, node, inputs, output_storage):
+        n_steps, *inputs = inputs
+        n_update = len(self.outputs) - (1 if self.is_while else 0)
+        carry, constant = inputs[:n_update], inputs[n_update:]
+        inner_fn = self.py_perform_fn
+
+        if self.is_while:
+            until = True
+            for i in range(n_steps):
+                *carry, until = inner_fn(*carry, *constant)
+                if until:
+                    break
+
+            if not until:  # no-break
+                if self.until_condition_failed == "raise":
+                    raise RuntimeError(
+                        f"Until condition in ScalarLoop {self.name} not reached!"
+                    )
+                elif self.until_condition_failed == "warn":
+                    warnings.warn(
+                        f"Until condition in ScalarLoop {self.name} not reached!",
+                        RuntimeWarning,
+                    )
+        else:
+            if n_steps < 0:
+                raise ValueError("ScalarLoop does not have a termination condition.")
+            for i in range(n_steps):
+                carry = inner_fn(*carry, *constant)
+
+        for storage, out_val in zip(output_storage, carry):
+            storage[0] = out_val
+
+    @property
+    def c_code_template(self):
+        from pytensor.link.c.interface import CLinkerType
+
+        if hasattr(self, "_c_code"):
+            return self._c_code
+
+        fgraph = self.fgraph
+
+        # The first input is `n_steps` so we skip it in the mapping dictionary
+        n_update = len(self.outputs) - (1 if self.is_while else 0)
+        carry_subd = {
+            c: f"%(i{int(i)})s" for i, c in enumerate(fgraph.inputs[:n_update], start=1)
+        }
+        constant_subd = {
+            c: f"%(i{int(i)})s"
+            for i, c in enumerate(fgraph.inputs[n_update:], start=n_update + 1)
+        }
+        update_subd = {
+            u: f"%(o{int(i)})s" for i, u in enumerate(fgraph.outputs[:n_update])
+        }
+        until_subd = {u: "until" for u in fgraph.outputs[n_update:]}
+        subd = {**carry_subd, **constant_subd, **update_subd, **until_subd}
+
+        for var in fgraph.variables:
+            if var.owner is None:
+                if var not in self.fgraph.inputs:
+                    # This is an orphan
+                    if isinstance(var, Constant) and isinstance(var.type, CLinkerType):
+                        subd[var] = var.type.c_literal(var.data)
+                    else:
+                        raise ValueError(
+                            "All orphans in the fgraph to ScalarLoop must"
+                            " be Constant, CLinkerType instances."
+                        )
+            elif any(i.dtype == "float16" for i in var.owner.inputs) or any(
+                o.dtype == "float16" for o in var.owner.outputs
+            ):
+                # flag for elemwise ops to check.
+                self.inner_float16 = True
```

Comment:
Here is why I brought up half-precision (float16). There's no test suite for it, but these checks are here, so they shouldn't have any test coverage

Reply:
Ah... I can add one, but I imagine everything will just work out fine.

Reply:
Yeah I figured. I'm nitpicking at this point.

Reply:
Now the inner tests use float16 instead of float32, which should cover it!